mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -0,0 +1,152 @@
+                service = BlobClient.from_blob_url(blob_url=signed_write_uri, credential=None)
+                service.commit_block_list(uploading_block_list)
+        except Exception as err:
+            raise MlflowException(err)
+
+    def _azure_download_file(self, credentials, local_path):
+        signed_read_uri = credentials.signed_uri
+        service = BlobClient.from_blob_url(blob_url=signed_read_uri, credential=None)
+        try:
+            with open(local_path, "wb") as output_file:
+                blob = service.download_blob()
+                output_file.write(blob.readall())
+        except Exception as err:
+            raise MlflowException(err)
+
+    def _aws_upload_file(self, credentials, local_file):
+        pass
+
+    def _aws_download_file(self, credentials, local_path):
+        pass
+
+    def _upload_to_cloud(self, cloud_credentials, local_file, artifact_path):
+        if cloud_credentials.credentials.type == 1:
+            self._azure_upload_file(cloud_credentials.credentials, local_file, artifact_path)
+        else:
+            raise MlflowException('Not implemented yet')
+
+    def _download_from_cloud(self, cloud_credentials, local_path):
+        if cloud_credentials.credentials.type == 1:
+            self._azure_download_file(cloud_credentials.credentials, local_path)
+        else:
+            raise MlflowException('Not implemented yet')
+
+    def log_artifact(self, local_file, artifact_path=None):
+        basename = os.path.basename(local_file)
+        artifact_path = artifact_path or ""
+        artifact_path = os.path.join(artifact_path, basename)
+        write_credentials = self._get_write_credentials(self.run_id, artifact_path)
+        self._upload_to_cloud(write_credentials, local_file, artifact_path)
+
+    def log_artifacts(self, local_dir, artifact_path=None):
+        artifact_path = artifact_path or ''
+        for (dirpath, _, filenames) in os.walk(local_dir):
+            artifact_subdir = artifact_path
+            if dirpath != local_dir:
+                rel_path = os.path.relpath(dirpath, local_dir)
+                rel_path = relative_path_to_artifact_path(rel_path)
+                artifact_subdir = os.path.join(artifact_path, rel_path)
+            for name in filenames:
+                file_path = os.path.join(dirpath, name)
+                self.log_artifact(file_path, artifact_subdir)
+
+    def list_artifacts(self, path=None):
+        json_body = message_to_json(ListArtifacts(run_id=self.run_id, path=path))
+        artifact_list = self._call_endpoint(MlflowService, ListArtifacts, json_body).files
+        # If `path` is a file, ListArtifacts returns a single list element with the
+        # same name as `path`. The list_artifacts API expects us to return an empty list in this
+        # case, so we do so here.
+        if len(artifact_list) == 1 and artifact_list[0].path == path:
+            return []
```

Comment:
Great catch!

Edit: Saw this logic is used in `dbfs_artifact_repo` as well - unfortunately, it looks like it's a bit brittle. To test it, I created a file called `foo` (it was created inside a directory also called `arty/foo`, but that's not important). Then, I ran the following code:

```
In[0]: client.list_artifacts(run_id, "arty/foo/foo")
Out[0]: []

In[1]: client.list_artifacts(run_id, "arty/foo/foo////")
Out[1]: [<FileInfo: file_size=5, is_dir=False, path='arty/foo/foo'>]
```

Looks like those trailing slashes break the path comparison. However, it's a bit strange that `ListArtifacts` returns the *file* path `arty/foo/foo` for an input of `arty/foo/foo/////`, since this output path is a file (for files, `test/` and `test//` are not equivalent...) *and* `arty/foo/foo/////` is not a prefix of `arty/foo/foo`. I think we should file a bug report ticket for the ListArtifacts API for this case.

Reply:
This makes sense. To fix this, does it make sense to perform a regex check, since if a users wants to download a single file, the format will :probably: end with a .txt, .py, .pkl to name a few. This could be a restriction if a single file needs to be downloaded.